app/api/currentuser/export_data.go

```diff
@@ -0,0 +1,200 @@
+package currentuser
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+)
+
+// swagger:operation GET /current-user/export-data users currentUserDataExport
+// ---
+// summary: Export the current user's data
+// description: >
+//   Returns a downloadable JSON file with all the current user's data.
+//   The content returned is just the dump of raw entries of each table related to the user
+//
+//     * `current_user` (from `users`): all attributes except `iVersion`
+//     * `sessions`, `refresh_token`: all attributes, but secrets replaced with “***”
+//     * `owned_groups`: `ID` and `sName` for every descendant of user’s `idGroupOwned`;
+//     * `joined_groups`: `ID` and `sName` for every ancestor of user’s `idGroupSelf`;
+//     * `users_answers`: all attributes;
+//     * `users_items`: all attributes except `iVersion`;
+//     * `groups_attempts`: the user's or his teams' attempts, all attributes except `iVersion`;
+//     * `groups_groups`: where the user’s `idGroupSelf` is the `idGroupChild`, all attributes except `iVersion` + `groups.sName`.
+// produces:
+//   - application/json
+// responses:
+//   "200":
+//     description: The returned data dump file
+//     schema:
+//       type: file
+//   "401":
+//     "$ref": "#/responses/unauthorizedResponse"
+//   "500":
+//     "$ref": "#/responses/internalErrorResponse"
+func (srv *Service) exportData(w http.ResponseWriter, r *http.Request) service.APIError {
+	user := srv.GetUser(r)
+
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.Header().Set("Content-Disposition", "attachment; filename=user_data.json")
```

Comment:
Unfortunately we cannot override headers after WriteHeader() has been called. That's a drawback of the JSON streaming. In case of an error, we respon with a malformed JSON ending with ```{"success":false,"error":"Unexpected error"}```. Don't know if we can do better.

Reply:
Ok, I didn't get that you were streaming JSON. Yeah the error case is a bit strange. I would expect test scenarios "documenting" this by example.

If we want to do better: 
* when ` w.Write(..)` fails, there is no much we could do, but then we could stop directly and the error message is useless but couldn't be written as well anyway... so maybe we can leave it as it is for now.
* when db fails, maybe we could just write something like the string "error fetching data" and continuing. It would allow to output a kind of valid output even if one query fails. Then the question may be "do we return 2xx in this case still?".

BTW there are no robustness case while there should be as well for the permission checks.

Reply:
> when db fails, maybe we could just write something like the string "error fetching data" and continuing

Reply:
> BTW there are no robustness case while there should be as well for the permission checks.

There are no permission checks in this service. The only check is performed by the user middleware and we already have tests for it.

Reply:
> I don't think we can continue after the DB failed. Not sure it makes sense at all. If the DB failed once it will fail on all the next queries. Also, not sure there is a reason to try output valid output when the data is corrupted (partially missing).

If the request is not correct (because some attributes were removed of whatever) you'll have the error just for one. You prefer having as output `{"current_user:{"success":false,"error":"Unexpected error"}`?

Reply:
The probability of the case where some attributes have been removed right between getColumnsList() and the query seems to be negligible. For me, `{"current_user":{"success":false,"error":"Unexpected error"}` is okay, at least it is not better than `{"current_user":{"success":false,"error":"Unexpected error"},"groups_groups":[...],...}`

Reply:
ok. Still require some tests then.

Reply:
added more tests